### PR TITLE
Fix the read-under-write behavior of register file

### PR DESCRIPTION
**Pull request by lsk567**
This PR uses the solution from @agrn to fix the read-under-write behavior of the register file.
A corresponding test case has been added to the unit test.

### CODE REVIEW

File: src/main/scala/Core/RegisterFile.scala [hunk 1]
```diff
@@ -61,19 +61,22 @@ class RegisterFile(val threads: Int, val readPorts: Int = 2, val writePorts: Int
       0.U
     )
 
+    // FIXME: Shaokai: Looks suspicious.
```
Comment (edwardcwang) [inline src/main/scala/Core/RegisterFile.scala:64]: ?

**Review by edwardcwang — COMMENTED** [review]
Thanks for your hard work debugging this. Would it be possible for you to add some more context to the changes, particularly w.r.t. the tests? Thank you!

File: src/main/scala/Core/RegisterFile.scala (hunk 1 above)
Comment (lsk567) [reply src/main/scala/Core/RegisterFile.scala]: Sorry, this legacy comment was left here when I looked at the original RegisterFile code. Thanks for catching this...

**Comment by lsk567** [comment]
The issue was that when the regfile is both read and written to in the same cycle (read-under-write), the original implementation does not check whether `enable` is set to true before returning the write data (written in the same cycle) to the read port.

The original unit test for read-under-write assumes that `enable` is set to true. The added test case checks that write data is **not** returned to the read port when `enable` is false.

**Pull request merged**